### PR TITLE
wasm `save_file`

### PR DESCRIPTION
**Pull request by Easyoakland**
Solving https://github.com/PolyMeilex/rfd/issues/88.

Create new `rfd::FileHandle::write(data: Box<[u8]>)` which should be able to be implemented for all targets including Wasm.

Currently only implemented for Wasm target and everything is in one function. It should almost certainly be be moved out to `WasmDialog`

I won't bother finishing this if it's indicated that this is an unwanted way to solve the issue, so feedback is welcome.

The other solution indicated appears to be
>There is also a chrome specific API that lets you open a file save dialog and get a file handle that way, this aligns with how paths work on native so we would just have to wrap those in a new type.

but that sounds like it won't work on Firefox.

### CODE REVIEW

**Comment by PolyMeilex** [comment]
So the way I see this being used would be:

On native:
- Call to save dialog would return FileHandle(Path) alike struct, just like pick dialog does now
- Users can either extract the raw path or call save method, that does the writing

On web:
- Call to save dialog would return fake/empty FileHandle alike struct, the dialog would not be opened
- When calling save on the handle the dialog would pops up and users could download the file that way.

Does this sound like a reasonable usage of this POC?

**Comment by Easyoakland** [comment]
> Call to save dialog would return FileHandle(Path) alike struct, just like pick dialog does now

I think that is already how it works as well in [save_file](https://docs.rs/rfd/latest/rfd/struct.AsyncFileDialog.html#method.save_file). Did you mean `save_file` when you wrote `save_dialog`?

> Users can either extract the raw path or call save method, that does the writing

Yes, though I named it `write` to match the [`read`](https://docs.rs/rfd/latest/rfd/struct.FileHandle.html#method.read) method.

> Call to save dialog would return fake/empty FileHandle alike struct, the dialog would not be opened

Yes. [`FileHandle::save_file`](https://docs.rs/rfd/latest/rfd/struct.FileDialog.html#method.save_file)

> When calling save on the handle the dialog would pops up and users could download the file that way.

Yes.

> I have to take a look at those APIs, to check if in the future that would do the trick there as well.

Sure.

> Does this sound like a reasonable usage of this POC?

Yes.

**Comment by PolyMeilex** [comment]
> I think that is already how it works as well in [save_file](https://docs.rs/rfd/latest/rfd/struct.AsyncFileDialog.html#method.save_file)

Yep, I forgot that I already did that

In that case this sounds awesome to me, I don't have any concerns.

**Comment by Easyoakland** [comment]
Done for native ([0345e44](https://github.com/PolyMeilex/rfd/pull/134/commits/0345e44b9c49bc2a9508e843a93fffbc96a597a2))

File: src/file_handle/native.rs [hunk 1]
```diff
@@ -57,6 +57,55 @@ impl Future for Reader {
...
+        }));
+
+        {
+            let path = path.to_owned();
+            let state = state.clone();
+            std::thread::Builder::new()
+                .name("rfd_file_write".into())
+                .spawn(move || {
+                    let mut state = state.lock().unwrap();
+                    let _res = std::fs::write(path, state.bytes.take().unwrap());
```
Comment (PolyMeilex) [inline src/file_handle/native.rs:83]: Wouldn't holding the lock here during the whole write operation cause calls to poll to block?

Comment (Easyoakland) [reply src/file_handle/native.rs]: Fixed in [6284197](https://github.com/PolyMeilex/rfd/pull/134/commits/62841970485454a8c2da21e0d36ba2f526917d1b)

File: src/file_handle/web.rs [hunk 2]
```diff
@@ -1,15 +1,31 @@
...
+    /// Wrap a [`web_sys::File`] for reading. Use with [`FileHandle::read`]
     pub fn wrap(file: web_sys::File) -> Self {
-        Self(file)
+        Self(WasmFileHandleKind::Readable(file))
+    }
+
+    /// Create a dummy `FileHandle`. Use with [`FileHandle::write`].
+    pub fn writable(dialog: FileDialog) -> Self {
+        FileHandle(WasmFileHandleKind::Writable(dialog))
     }
```
Comment (PolyMeilex) [inline src/file_handle/web.rs:22]: What's the purpose of this one?

tbh I'm not even sure why I made regular wrap public, so if there is no specific reason I would refrain from adding another, especially platform specific one

**Review by PolyMeilex — APPROVED** [review]
In general this looks ok, I will do some testing with it soon'ish, and it should be good to go

File: src/file_handle/web.rs (hunk 2 above)
Comment (Easyoakland) [reply src/file_handle/web.rs:22]: It is used here `src\backend\wasm\file_dialog.rs:13:32`.
I made it `pub` to match the other `wrap`.
I can make it `pub(crate)`

Comment (PolyMeilex) [reply src/file_handle/web.rs:22]: pub(crate) would be good, I think I will change regular wraps to that as well, unless I find a reason why those are public :smile:

Comment (Easyoakland) [reply src/file_handle/web.rs]: Fix: [c0d0ba5](https://github.com/PolyMeilex/rfd/pull/134/commits/c0d0ba5bb2ddc7f1c291c6c7c2d3c7ebf7732f8d)

Obviously PR is now a major version change if it wasn't already.

**Comment by PolyMeilex** [comment]
Thanks!

**Comment by ZeWaka** [comment]
Thank you *so much* for this, actually made my day not having to implement this myself 😅

**Pull request merged**